src/librustc_mir/build/matches/mod.rs

```diff
@@ -38,6 +38,10 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                       -> BlockAnd<()> {
         let discriminant_place = unpack!(block = self.as_place(block, discriminant));
 
+        let arms: Vec<Arm<'tcx>> = arms.into_iter().filter(|arm|
+            arm.patterns.iter().any(|pat| !pat.is_unreachable())
```

Comment:
given that we didn't notice that the condition here was wrong, can we make a test targeting that particular case?

Reply:
Yep, will do!

Reply:
I've added a MIR test that differentiates between these two cases.